File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/support/tree/support/Markup.java
package net.nemerosa.ontrack.model.support.tree.support;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.HashMap;
import java.util.Map;

@Data
@AllArgsConstructor(access = AccessLevel.PROTECTED)
public class Markup {

    private final String type;
    private final String text;
    private final Map<String, String> attributes;

    public static Markup text(String text) {
        return new Markup(null, text, null);
    }

    public static Markup of(String type) {
        return new Markup(type, null, new HashMap<String, String>());
    }

    public static Markup of(String type, Map<String, String> attributes) {
        return of(type).attrs(attributes);
    }

    public Markup attrs(Map<String, String> attributes) {
        this.attributes.putAll(attributes);
        return this;
    }

    public Markup attr(String name, String value) {
        attributes.put(name, value);
        return this;
    }

    @JsonIgnore
    public boolean isOnlyText() {
        return type == null;
    }
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/support/tree/support/Markup.kt
package net.nemerosa.ontrack.model.support.tree.support

import com.fasterxml.jackson.annotation.JsonIgnore

data class Markup(
    val type: String?,
    val text: String?,
    val attributes: Map<String, String?>?,
) {

    fun attrs(attributes: Map<String, String?>) = copy(attributes = (this.attributes ?: emptyMap()) + attributes)

    fun attr(name: String, value: String?) = copy(
        attributes = (attributes ?: emptyMap()) + (name to value)
    )

    @get:JsonIgnore
    val isOnlyText: Boolean = type == null

    companion object {
        @JvmStatic
        fun text(text: String?): Markup {
            return Markup(null, text, null)
        }

        @JvmStatic
        fun of(type: String?): Markup {
            return Markup(type, null, emptyMap())
        }

        @JvmStatic
        fun of(type: String?, attributes: Map<String, String?>): Markup {
            return of(type).attrs(attributes)
        }
    }
}
